src/sentry/event_hashing.py

```diff
@@ -11,13 +11,292 @@
 DEFAULT_FINGERPRINT_VALUES = frozenset(['{{ default }}', '{{default}}'])
 
 
-def md5_from_hash(hash_bits):
+DEFAULT_HINTS = {
+    'salt': 'a static salt',
+}
+
+
+def _calculate_contributes(values):
+    for value in values or ():
+        if not isinstance(value, GroupingComponent) or value.contributes:
+            return True
+    return False
+
+
+class GroupingComponent(object):
+    """A grouping component is a recursive structure that is flattened
+    into components to make a hash for grouping purposes.
+    """
+
+    def __init__(self, id, hint=None, contributes=None, values=None):
+        self.id = id
+        if hint is None:
+            hint = DEFAULT_HINTS.get(id)
```

Comment:
I wanted to refactor this further by instead of passing `id` i pass instances of preconfigured GroupTypes or something. Then it gets clearer. On the UI branch I also now derive names from this so one needs to know that the IDs are not just entirely arbitrary. I think that will clarify this a little bit.

So it might turn into something like this:

```python
salt = GroupType(id='salt', hint='a static salt')
GroupingComponent(ty=salt, values=['...'])
```